docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -0,0 +1,544 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Quarkus - CDI Integration Guide
+
+include::./attributes.adoc[]
+:numbered:
+:sectnums:
+:sectnumlevels: 4
+:toc:
+:toclevels: 2
+
+The ArC container is bootstrapped at build time.
+The downside of this approach is that CDI Portable Extensions cannot be supported.
+Nevertheless, the functionality can be achieved using the Quarkus-specific extensions API.
+
+The container is bootstrapped in multiple phases.
+From a high level perspective these phases go as follows:
+
+1. Initialization
+2. Bean discovery 
+3. Registration of synthetic components
+4. Validation
+
+In the _initialization_ phase the preparatory work is being carried out and custom contexts are registered.
+_Bean discovery_ is then the process where the container analyzes all application classes, identifies beans and wires them all together based on the provided metadata.
+Subsequently, the extensions can register _synthetic components_.
+Attributes of these components are fully contolled by the extensions, i.e. are not derived from an existing class.
+Finally, the _deployment is validated_.
+For example, the container validates every injection point in the application and fails the build if there is no bean that satisfies the given required type and qualifiers.
+
+TIP: You can see more information about the bootstrap by enabling additional logging. Simply run the Maven build with `-X,--debug` and grep lines that contain `io.quarkus.arc`. In the <<cdi-reference.adoc#dev-mode,development mode>>, you can use `quarkus.log.category."io.quarkus.arc.processor".level=DEBUG` and two special endpoints are also registered automatically to provide some basic debug info in the JSON format.
+
+Quarkus build steps can produce and consume various build items and hook into each phase.
+In the following sections we will describe all the relevant build items and common scenarios.
+
+[IMPORTANT]
+.Metadata Sources
+====
+Classes and annotations are the primary source of bean-level metadata.
+However, extensions can add, remove or transform the metadata at certain stages of the bootstrap.
+Moreover, extensions can also register <<synthetic_beans,synthetic components>>. 
+This is an important aspect to realize when integrating CDI components in Quarkus.
```

Comment:
It would be helpful to explain in greater details why it's important (because some beans that you might expect might not be there anymore at the end of the process, more are added, etc.) and maybe provide some examples to show concretely what might happen.

Reply:
That's a good idea but I'd like to keep this part of the docs concise. What about this: _"Quarkus extensions can turn classes, that would be otherwise ignored, into beans and vice versa. For example, a class that declares a `@Scheduled` method is always registered as a bean even if it is not annotated with a bean defining annotation and would be normally ignored."_


In fact, it's probably a good topic for a blogpost or something like that. The list of examples of "what might happen" could be very long...

Reply:
I agree that it would be a good material for a blog post (and I intend to document some of my extension writing experience in such blogs). 
However, when looking at reference doc, I'd also expect it to be as complete as possible so that I don't have to look at lots of different places to find the information I need. The example you give would probably be good enough if we don't want to get into too many details here.

Reply:
> I'd also expect it to be as complete as possible so that I don't have to look at lots of different places to find the information I need.

It's always a question of balance. My concern is that "too exhaustive" docs may scare people ;-).

We can always add more examples.